File: src/orchestrator/api/schemas/clarifications.py
"""Clarification API schemas."""

from datetime import datetime
from typing import Literal

from pydantic import model_validator

from orchestrator.api.schemas.base import ApiModel


class ClarificationQuestionSchema(ApiModel):
    """A question from the builder needing human input."""

    id: str
    question: str
    context: str
    options: list[str] = []
    question_type: Literal["single_select", "multi_select", "free_text", "number"] = "single_select"
    allow_other: bool = True
    required: bool = True
    min: float | None = None
    max: float | None = None
    placeholder: str | None = None

    @model_validator(mode="after")
    def validate_options_for_type(self) -> "ClarificationQuestionSchema":
        if self.question_type in ("single_select", "multi_select"):
            if not self.options:
                raise ValueError(
                    f"'options' must be non-empty for question_type={self.question_type!r}"
                )
        else:  # free_text, number
            if self.options:
                raise ValueError(
                    f"'options' must be empty for question_type={self.question_type!r}"
                )
        if self.min is not None and self.max is not None and self.min > self.max:
            raise ValueError("'min' must be <= 'max'")
        return self


class ClarificationAnswerSchema(ApiModel):
    """Human's answer to a clarification question."""

    question_id: str
    selected_option: str | None = None
    free_text: str | None = None
    selected_options: list[str] | None = None
    skipped: bool = False
    skip_reason: str | None = None


class CreateClarificationRequest(ApiModel):
    """Request to create a clarification from the builder."""

    questions: list[ClarificationQuestionSchema]


class ClarificationRequestResponse(ApiModel):
    """Response containing a clarification request."""

    id: str
    run_id: str
    task_id: str
    attempt_num: int
    questions: list[ClarificationQuestionSchema]
    created_at: datetime
    responded_at: datetime | None = None


class RespondToClarificationRequest(ApiModel):
    """Request to submit answers to clarification questions."""

    answers: list[ClarificationAnswerSchema]
    skipped: bool = False
    skip_reason: str | None = None


class ClarificationHistoryItem(ApiModel):
    """A single clarification round (request + optional response)."""

    request: ClarificationRequestResponse
    response: RespondToClarificationRequest | None = None


class ClarificationHistoryResponse(ApiModel):
    """All clarification rounds for a task."""

    items: list[ClarificationHistoryItem]


class PendingActionSchema(ApiModel):
    """A pending user action (clarification or approval)."""

    task_id: str
    step_id: str
    action_type: str  # "clarification" | "approval"
    clarification_request: ClarificationRequestResponse | None = None
    summary_artifact: str | None = None
    approval_prompt: str | None = None
    is_gate_approval: bool = False

File: ui/src/types/clarifications.ts
export interface ClarificationQuestion {
  id: string;
  question: string;
  context: string;
  options: string[];
  question_type: 'single_select' | 'multi_select' | 'free_text' | 'number';
  allow_other: boolean;
  required: boolean;
  min?: number | null;
  max?: number | null;
  placeholder?: string | null;
}

export interface ClarificationAnswer {
  question_id: string;
  selected_option: string | null;
  free_text: string | null;
  selected_options?: string[];
  skipped?: boolean;
  skip_reason?: string | null;
}

export interface ClarificationRequest {
  id: string;
  run_id: string;
  task_id: string;
  attempt_num: number;
  questions: ClarificationQuestion[];
  created_at: string;
  responded_at: string | null;
}

export interface RespondToClarificationRequest {
  answers: ClarificationAnswer[];
  skipped?: boolean;
  skip_reason?: string | null;
}

export interface ClarificationResponse {
  request_id?: string;
  answers: ClarificationAnswer[];
  responded_at?: string;
  skipped?: boolean;
  skip_reason?: string | null;
}

export interface ClarificationHistoryItem {
  request: ClarificationRequest;
  response: ClarificationResponse | null;
}

export interface ClarificationHistoryResponse {
  items: ClarificationHistoryItem[];
}

export interface PendingAction {
  task_id: string;
  step_id: string;
  action_type: 'clarification' | 'approval';
  clarification_request: ClarificationRequest | null;
  summary_artifact: string | null;
  approval_prompt: string | null;
  is_gate_approval: boolean;
}

export interface ApproveTaskRequest {
  comment?: string;
}

export interface RejectTaskRequest {
  reason?: string;
}

export interface ForceAcceptTaskRequest {
  comment?: string;
}
